account_statement_import_txt_xlsx/models/account_statement_import_sheet_parser.py

```diff
@@ -0,0 +1,378 @@
+                "amount": amount,
+                "currency": currency,
+                "original_amount": original_amount,
+                "original_currency": original_currency,
+            }
+            if balance is not None:
+                line["balance"] = balance
+            if transaction_id is not None:
+                line["transaction_id"] = transaction_id
+            if description is not None:
+                line["description"] = description
+            if notes is not None:
+                line["notes"] = notes
+            if reference is not None:
+                line["reference"] = reference
+            if partner_name is not None:
+                line["partner_name"] = partner_name
+            if bank_name is not None:
+                line["bank_name"] = bank_name
+            if bank_account is not None:
+                line["bank_account"] = bank_account
+            lines.append(line)
+        return lines
+
+    @api.model
+    def _convert_line_to_transactions(self, line):
+        """Hook for extension"""
+        timestamp = line["timestamp"]
+        amount = line["amount"]
+        currency = line["currency"]
+        original_amount = line["original_amount"]
+        original_currency = line["original_currency"]
+        transaction_id = line.get("transaction_id")
+        description = line.get("description")
+        notes = line.get("notes")
+        reference = line.get("reference")
+        partner_name = line.get("partner_name")
+        bank_name = line.get("bank_name")
+        bank_account = line.get("bank_account")
+
+        transaction = {
+            "date": timestamp,
+            "amount": str(amount),
+        }
+        if currency != original_currency:
+            original_currency = self.env["res.currency"].search(
+                [("name", "=", original_currency)],
+                limit=1,
+            )
+            if original_currency:
+                transaction.update(
+                    {
+                        "amount_currency": str(original_amount),
+                        "currency_id": original_currency.id,
+                    }
+                )
+
+        if transaction_id:
+            transaction["unique_import_id"] = "{}-{}".format(
+                transaction_id,
+                int(timestamp.timestamp()),
+            )
+
+        transaction["name"] = description or _("N/A")
+        if reference:
+            transaction["ref"] = reference
+
+        note = ""
+        if bank_name:
+            note += _("Bank: %s; ") % (bank_name,)
+        if bank_account:
+            note += _("Account: %s; ") % (bank_account,)
+        if transaction_id:
+            note += _("Transaction ID: %s; ") % (transaction_id,)
+        if note and notes:
+            note = "{}\n{}".format(note, note.strip())
+        elif note:
+            note = note.strip()
+        if note:
+            transaction["payment_ref"] = note
```

Comment:
replace if with:
```transaction["payment_ref"] = reference or description or note or transaction.get("unique_import_id") or "/"```

"payment_ref" must be present. The order of choice is open for discussion. In practical examples I've seen, often there's only a description line.  Which one to choose if note, reference and description isn't an obvious pick.

Here's the rationale for my choice: let the user pick from actual data first, then provide generated data (on which the user has less control) as fallback.

Reply:
Thanks, merged with https://github.com/PNLUG/bank-statement-import/pull/4.